### PR TITLE
Revert the output pipe in the DuplexStreamPipeAdapter

### LINKED ISSUE #11560: Network errors when refreshing an app repeatedly
I created a blazorserverside app using the latest sdk from core-sdk

dotnet new blazorserverside
dotnet run
Open chrome
Press F5 repeatedly at some point a request will fail and the UI will change.

A8mjiYKM8d

### PR DESCRIPTION
**Pull request by davidfowl**
- Don't complete the connection pipe in Http2FrameWriter 
- Replace the output pipe only
- Updated tests

Fixes #11560

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 1]
```diff
@@ -14,36 +15,110 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
...
+                                                readerScheduler: PipeScheduler.Inline,
+                                                writerScheduler: PipeScheduler.Inline,
+                                                pauseWriterThreshold: 1,
+                                                resumeWriterThreshold: 1,
+                                                minimumSegmentSize: writerOptions.MinimumBufferSize,
+                                                useSynchronizationContext: false);
+
             Input = PipeReader.Create(Stream, readerOptions);
-            Output = PipeWriter.Create(Stream, writerOptions);
+            Output = new Pipe(outputOptions);
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: This probably needs a comment on why we are doing this.

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 2]
```diff
@@ -14,36 +15,110 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
...
         }
 
+        public ILogger Log { get; set; }
+
         public TStream Stream { get; }
 
+        private Pipe Output { get; }
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: nit: why inconsistencies on private and public for input/output.

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 3]
```diff
@@ -14,36 +15,110 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
...
+            try
+            {
+                while (true)
+                {
+                    var result = await Output.Reader.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    try
+                    {
+                        if (buffer.IsEmpty)
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: I think you should check result.IsCompleted even if the buffer isn't empty.

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: This code isn't new, why make tweaks to it at this point?

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: It also doesn't make sense to check IsCompleted for non empty buffers. That can result in truncation.

**Review by jkotalik — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 4]
```diff
@@ -80,39 +80,7 @@ private async Task WriteOutputAsync()
...
-                                await Stream.WriteAsync(memory);
-                            }
-                        }
-                    }
-                    finally
-                    {
-                        _output.Reader.AdvanceTo(buffer.End);
-                    }
-                }
+                await _output.Reader.CopyToAsync(Stream);
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: 🌶

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 5]
```diff
@@ -14,36 +15,114 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
...
+                    }
+                    finally
+                    {
+                        _output.Reader.AdvanceTo(buffer.End);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log?.LogError(0, ex, $"{GetType().Name}.{nameof(WriteOutputAsync)}");
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: Given that both LibuvConnection and SocketConnection will not complete their PipeReaders with an error unless it was completely unexpected, it's tempting to make this a critical log. That's unless we're worried about something like SslStream throwing an Exception in the middle of WriteAsync despite there being no transport issue.

Maybe that is something we should be worried about... I'm just extra worried about swallowing errors now.

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs [hunk 6]
```diff
@@ -14,36 +15,114 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
...
+        public DuplexPipeStreamAdapter(IDuplexPipe duplexPipe, StreamPipeReaderOptions readerOptions, StreamPipeWriterOptions writerOptions, Func<Stream, TStream> createStream) :
+            base(duplexPipe.Input, duplexPipe.Output)
         {
             Stream = createStream(this);
+
+            var outputOptions = new PipeOptions(pool: writerOptions.Pool,
+                                                readerScheduler: PipeScheduler.Inline,
+                                                writerScheduler: PipeScheduler.Inline,
+                                                pauseWriterThreshold: 1,
+                                                resumeWriterThreshold: 1,
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: We should probably use the defaults here like we did before to reduce thrashing:

https://github.com/aspnet/AspNetCore/blob/release/3.0-preview6/src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs#L67-L76

```c#
        internal PipeOptions AdaptedOutputPipeOptions => new PipeOptions
        (
            pool: MemoryPool,
            readerScheduler: PipeScheduler.Inline,
            writerScheduler: PipeScheduler.Inline,
            pauseWriterThreshold: _context.ServiceContext.ServerOptions.Limits.MaxResponseBufferSize ?? 0,
            resumeWriterThreshold: _context.ServiceContext.ServerOptions.Limits.MaxResponseBufferSize ?? 0,
            useSynchronizationContext: false,
            minimumSegmentSize: MemoryPool.GetMinimumSegmentSize()
        );
```

**Review by halter73 — APPROVED** [review]
Please fix the pauseWriterThreshold and resumeWriterThreshold before merging though.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs [hunk 7]
```diff
@@ -89,7 +89,6 @@ public void Complete()
 
                 _completed = true;
                 _connectionOutputFlowControl.Abort();
-                _outputWriter.Complete();
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs]: 👍

**Comment by halter73** [comment]
Maybe trying out the lower thresholds are OK for a preview. It does seem a little unecsary to try out new options now if we're removing it completely in the next preview though.

**Comment by davidfowl** [comment]
> Maybe trying out the lower thresholds are OK for a preview. It does seem a little unecsary to try out new options now if we're removing it completely in the next preview though.

I'd rather leave it as it so its as passthrough as possible. Also the fact that it is temporary and the fact that those options aren't readily available (they need to be flowed from the UseHttps call) is just more churn. If you do feel strongly about changing those limits back to the Kestrel limits.

**Comment by halter73** [comment]
> If you do feel strongly about changing those limits back to the Kestrel limits, I'll do it.

I'm fine with the way it is since it's temporary.

File: src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs (hunk 5 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: I'm going to leave it at error for now. I do agree that swallowing errors is a concern, but what else should we do. Debug.Assert? :trollface:

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Middleware/Internal/DuplexPipeStreamAdapter.cs]: David already passed in a logger here. Critical logs are the new, much-improved Debug.Assert. Any critical log should cause the majority of our tests to fail.

**Comment by davidfowl** [comment]
> I'm fine with the way it is since it's temporary. Otherwise, I'd at least ask for benchmarks.

Maybe it's reasonable to leave the defaults then.

**Comment by davidfowl** [comment]
> I'm fine with the way it is since it's temporary. Otherwise, I'd at least ask for benchmarks.

I just realized though, you'll very likely never hit those limits since we always consume the entire buffer and the scheduling is inline. It means call to write will yield the await, write to the Stream then call Advance). It'll only matter when we're experiencing back pressure from the transport itself (which is the effect we want, complete passthrough).

**Comment by halter73** [comment]
> I just realized though, you'll very likely never hit those limits since we always consume the entire buffer and the scheduling is inline. It means call to write will yield the await, write to the Stream then call Advance). It'll only matter when we're experiencing back pressure from the transport itself (which is the effect we want, complete passthrough).

This has always been my assumption. That's why I'm not *too* worried. I like validating my assumptions though.

**Pull request merged**